Refactor actor lookup order for HTTP URIs

pfefferle opened the pull request:
Moved the check for locally stored remote actors to occur after blog actor detection in class-actors.php. Updated class-query.php to handle blog actor ID for home/front page URLs before falling back to author ID lookup. This improves accuracy in actor resolution for blog and remote actors.

## Proposed changes:
- Remove `Remote_Actors::get_by_uri` lookup because it is not functional and slows down lookups.
- In query resolution, map home/front page URLs to BLOG_USER_ID before falling back to author ID resolution.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [X] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [X] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [X] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Improved actor resolution by prioritizing blog actor detection before remote actor checks and refining home page URL handling.

</details>

pfefferle committed:
Refactor actor lookup order for HTTP URIs

Moved the check for locally stored remote actors to occur after blog actor detection in class-actors.php. Updated class-query.php to handle blog actor ID for home/front page URLs before falling back to author ID lookup. This improves accuracy in actor resolution for blog and remote actors.

pfefferle committed:
Refactor blog actor check in Query class

Simplifies the logic for checking if the blog user type is disabled on home or front page. Now returns null immediately if the blog user type is disabled, improving code clarity.

pfefferle committed:
Add tests for home page in actor modes

Introduces tests for the get_activitypub_object method to verify correct behavior on the home page in both Actor and Actor-and-Blog modes. Ensures that no queries are made to the Remote_Actors table and validates the returned object.

pfefferle committed:
Skip home query for ActivityPub requests

Adds a parse_query action to set is_home to false when ActivityPub-specific query vars are present, preventing unnecessary database queries for the home page when only the blog actor is needed.

pfefferle committed:
Add global namespace to is_home and is_front_page

Prefixed is_home() and is_front_page() with a backslash to ensure global function resolution, preventing potential conflicts with local implementations.

matticbot committed:
Add changelog

pfefferle committed:
Remove local remote Actor lookup in get_user_id_by_uri

Eliminates the check for locally stored remote Actor in the get_user_id_by_uri method, streamlining the user lookup process and returning an error if the Actor is not found.

pfefferle committed:
Remove blog actor check from get_activitypub_object

Simplifies author ID resolution by removing the special case for blog actors on the home and front page. Updates related test to use fully qualified function calls for add_filter and remove_filter.

pfefferle committed:
Replace add_option with update_option in tests

Changed add_option to update_option for 'activitypub_actor_mode' in test cases to ensure the option is updated rather than added, improving test reliability when the option may already exist.

pfefferle committed:
Merge branch 'trunk' into improve-home-page-performance

obenland commented on tests/phpunit/tests/includes/class-test-query.php:
```diff
@@ -518,4 +518,77 @@ public function test_maybe_get_stamp_invalid_author() {
+			$message .= ' Found queries: ' . wp_json_encode( $actor_queries );
+		}
+
+		$this->assertNull( $object, 'Home page should return an object' );
```
One of these feels off. Should it be null or an object? If an object, do we expect a specific object that we can test for?

pfefferle replied:
It should be null, because the Blog-User is disabled in that case.

pfefferle replied:
I updated the description.

obenland commented on tests/phpunit/tests/includes/class-test-query.php:
```diff
@@ -518,4 +518,77 @@ public function test_maybe_get_stamp_invalid_author() {
+			$message .= ' Found queries: ' . wp_json_encode( $actor_queries );
+		}
+
+		$this->assertNotNull( $object, 'Home page should return an object' );
```
Same as above.

pfefferle replied:
It does not really matter what object it is, I want to simply proof that there is no DB lookup in this case.

obenland commented on includes/collection/class-actors.php:
```diff
@@ -222,13 +222,6 @@ public static function get_id_by_resource( $uri ) {
-				if ( ! \is_wp_error( $post ) ) {
-					return $post->ID;
-				}
-
```
The PR description says "Moved the check for locally stored remote actors to occur after blog actor detection" but that doesn't seem to be the case?

Is there a reason `maybe_get_virtual_object()` calls `Actors::get_by_various()` at all? could we not just return either the Blog or Application object based on whatever query parameters we need?

pfefferle replied:
> The PR description says "Moved the check for locally stored remote actors to occur after blog actor detection" but that doesn't seem to be the case?

Yes, my Copilot budget was expired, so it didn't catch up :) (I update the code).

The fallback to `Remote_Actors::get_by_uri( $uri )` is not functional, because it returns a Post-ID and we only check for User-IDs, so I removed it.

pfefferle replied:
> Is there a reason maybe_get_virtual_object() calls Actors::get_by_various() at all? could we not just return either the Blog or Application object based on whatever query parameters we need?

The Blog-User and the Application users have several Identifiers, so I would have to clone the complete code of `Actors::get_by_various()`.

For example:

Blog:

* `http://example.org/`
* `http://example.org/@example.org`
* `http://example.org/?author=0`

Application:

* `http://example.org/wp-json/activitypub/1.0/application`
* `http://example.org/@application`
* `http://example.org/?author=-1`

And because `is_home` is not very reliable (see Slack discussion), I decided to [remove it](https://github.com/Automattic/wordpress-activitypub/pull/2327/commits/6b31c175df7d75876396feed1d009f187978f303) and optimize the lookup instead. Happy for any advice!

obenland replied:
No I think that makes sense. Thanks!

pfefferle committed:
Clarify assertion message in Test_Query

Updated the assertion message in Test_Query to specify that the home page should return null because the Blog user is disabled, improving test clarity.
